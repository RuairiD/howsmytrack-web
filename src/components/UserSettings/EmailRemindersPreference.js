import React from "react";
import { useMutation } from "react-query";

import { Switch, Typography } from "antd";
import { Div } from "lemon-reset";
import axios from "axios";
import apiRoot from "../../apiRoot";

export type EmailRemindersPreferenceProps = {
    currentSendReminderEmails: boolean,
};

const UPDATE_SEND_REMINDER_EMAILS_MUTATION = `mutation UpdateSendReminderEmails($sendReminderEmails: Boolean!) {
    updateSendReminderEmails(sendReminderEmails: $sendReminderEmails) {
        success
    }
}`;

const EmailRemindersPreference = ({ currentSendReminderEmails }: EmailRemindersPreferenceProps) => {
    const sendReminderEmailsRequest = (checked) => (
        axios.post(`${apiRoot}/graphql/`, {
            query: UPDATE_SEND_REMINDER_EMAILS_MUTATION,
            variables: {
                sendReminderEmails: checked,
            },
        })
    );

    const [onSendReminderEmailsChange, { isLoading, data }] = useMutation(sendReminderEmailsRequest);

    return (
        <Div>
            <Typography.Text strong style={{ marginRight: "0.5em" }}>
                Email Reminders:
            </Typography.Text>
            <Switch
                defaultChecked={currentSendReminderEmails}
                onChange={onSendReminderEmailsChange}
                style={{ marginRight: "1em" }}
                loading={isLoading}
            />
            {data && data.success && (
                <Typography.Text type="secondary">
                    Saved
                </Typography.Text>
            )}
        </Div>
    );
};

export default EmailRemindersPreference;
