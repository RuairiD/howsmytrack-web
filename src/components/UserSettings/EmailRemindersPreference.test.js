import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import axios from "axios";
import { act } from "react-dom/test-utils";
import waitForExpect from "wait-for-expect";
import EmailRemindersPreference from "./EmailRemindersPreference";

jest.mock("axios");

describe("EmailRemindersPreference", () => {
    afterEach(() => {
        axios.post.mockRestore();
    });

    it("renders a labeled Switch", () => {
        const wrapper = shallow(
            <EmailRemindersPreference
                currentSendReminderEmails
            />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("submits a request to change the email reminders preference and shows success if the user toggles the switch", async () => {
        axios.post.mockImplementationOnce(() => Promise.resolve({
            data: {
                data: {
                    updateSendReminderEmails: {
                        success: true,
                    },
                },
            },
        }));

        const wrapper = mount(
            <EmailRemindersPreference
                currentSendReminderEmails
            />,
        );

        await act(async () => wrapper.find("Switch").get(0).props.onChange(false));

        expect(axios.post).toHaveBeenCalled();

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.text().includes("Saved")).toBeTruthy();
        });
    });

    it("does not show success indicator if request is unsuccessful.", async () => {
        axios.post.mockImplementationOnce(() => Promise.resolve({
            data: {
                data: {
                    updateSendReminderEmails: {
                        success: false,
                    },
                },
            },
        }));

        const wrapper = mount(
            <EmailRemindersPreference
                currentSendReminderEmails
            />,
        );

        await act(async () => wrapper.find("Switch").get(0).props.onChange(true));

        expect(axios.post).toHaveBeenCalled();

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.text().includes("Saved")).not.toBeTruthy();
        });
    });
});
