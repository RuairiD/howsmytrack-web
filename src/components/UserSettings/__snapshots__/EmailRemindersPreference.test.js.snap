// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmailRemindersPreference renders a labeled Switch 1`] = `
<Div
  className=""
>
  <Text
    strong={true}
    style={
      Object {
        "marginRight": "0.5em",
      }
    }
  >
    Email Reminders:
  </Text>
  <Switch
    defaultChecked={true}
    loading={false}
    onChange={[Function]}
    style={
      Object {
        "marginRight": "1em",
      }
    }
  />
</Div>
`;
